File: maui-client/Services/Protocol.cs
using System.Buffers.Binary;
using System.Text;

namespace BridgeToFreedom.Services;

/// <summary>
/// Wire protocol for the IP-tunnel build, mirroring the Go adapter/helper.
/// Format: [1B type][payload...]
/// PACKET payload = one raw IP packet.
/// PACKET_BATCH payload = repeated [2B lenBE][rawIPpacket].
/// </summary>
public static class Protocol
{
    // Control
    public const byte MsgHello       = 0x01;
    public const byte MsgHelloOK     = 0x02;
    public const byte MsgHelloErr    = 0x03;
    public const byte MsgPeerConn    = 0x04;
    public const byte MsgPeerGone    = 0x05;
    public const byte MsgSync        = 0x06;
    public const byte MsgPing        = 0xF0;
    public const byte MsgPong        = 0xF1;

    // Data
    public const byte MsgPacket      = 0x10;
    public const byte MsgPacketBatch = 0x11;

    public static string MsgName(byte type) => type switch
    {
        MsgHello       => "HELLO",
        MsgHelloOK     => "HELLO_OK",
        MsgHelloErr    => "HELLO_ERR",
        MsgPeerConn    => "PEER_CONN",
        MsgPeerGone    => "PEER_GONE",
        MsgSync        => "SYNC",
        MsgPing        => "PING",
        MsgPong        => "PONG",
        MsgPacket      => "PACKET",
        MsgPacketBatch => "PACKET_BATCH",
        _ => $"0x{type:X2}"
    };

    public static byte[] Encode(byte type, byte[]? payload = null)
    {
        var p = payload ?? [];
        var buf = new byte[1 + p.Length];
        buf[0] = type;
        if (p.Length > 0) p.CopyTo(buf, 1);
        return buf;
    }

    public static (byte Type, byte[] Payload) Decode(byte[] data)
    {
        if (data.Length < 1) throw new InvalidDataException("frame too short");
        var payload = data.Length > 1 ? data[1..] : [];
        return (data[0], payload);
    }

    public static byte[] EncodeHello(byte version, string token)
    {
        var t = Encoding.UTF8.GetBytes(token);
        var buf = new byte[1 + t.Length];
        buf[0] = version;
        t.CopyTo(buf, 1);
        return buf;
    }

    public static (string OwnId, string PeerId, string IamToken) DecodeHelloOK(byte[] payload)
    {
        int off = 0;
        var ownId = ReadLenPrefixed(payload, ref off);
        var peerId = ReadLenPrefixed(payload, ref off);
        var iamToken = ReadLenPrefixed(payload, ref off);
        return (ownId, peerId, iamToken);
    }

    public static (string PeerId, string IamToken) DecodePeerConn(byte[] payload)
    {
        int off = 0;
        var peerId = ReadLenPrefixed(payload, ref off);
        var iamToken = ReadLenPrefixed(payload, ref off);
        return (peerId, iamToken);
    }

    public static string DecodePong(byte[] payload)
    {
        int off = 0;
        return ReadLenPrefixed(payload, ref off);
    }

    /// <summary>
    /// Decodes a PACKET_BATCH payload, invoking <paramref name="onPacket"/> for each
    /// embedded IP packet. The slice handed to the callback is owned by <paramref name="payload"/>;
    /// copy if you need to retain it.
    /// </summary>
    public static int DecodePacketBatch(byte[] payload, Action<ArraySegment<byte>> onPacket)
    {
        int off = 0;
        int count = 0;
        while (off < payload.Length)
        {
            if (off + 2 > payload.Length)
                throw new InvalidDataException("PACKET_BATCH: truncated length prefix");
            var len = BinaryPrimitives.ReadUInt16BigEndian(payload.AsSpan(off));
            off += 2;
            if (off + len > payload.Length)
                throw new InvalidDataException("PACKET_BATCH: truncated packet");
            onPacket(new ArraySegment<byte>(payload, off, len));
            off += len;
            count++;
        }
        return count;
    }

    public static byte[] EncodePacketBatch(IReadOnlyList<byte[]> packets)
    {
        int total = 0;
        for (int i = 0; i < packets.Count; i++) total += 2 + packets[i].Length;
        var buf = new byte[total];
        int off = 0;
        for (int i = 0; i < packets.Count; i++)
        {
            var p = packets[i];
            BinaryPrimitives.WriteUInt16BigEndian(buf.AsSpan(off), (ushort)p.Length);
            off += 2;
            p.CopyTo(buf, off);
            off += p.Length;
        }
        return buf;
    }

    private static string ReadLenPrefixed(byte[] data, ref int off)
    {
        if (off + 2 > data.Length) throw new InvalidDataException("payload too short");
        var len = BinaryPrimitives.ReadUInt16BigEndian(data.AsSpan(off));
        off += 2;
        if (off + len > data.Length) throw new InvalidDataException("payload truncated");
        var s = Encoding.UTF8.GetString(data, off, len);
        off += len;
        return s;
    }
}

File: adapter-and-helper/internal/protocol/protocol.go
package protocol

import (
	"encoding/binary"
	"errors"
)

// Wire format: [1B type][payload...]
//
// Both sides are symmetric: they read raw IP packets from a local TUN device,
// wrap each one in a PACKET frame, ship it through the WebSocket bridge to the
// peer, and the peer writes it to its own TUN. The OS TCP/IP stack on each end
// handles fragmentation, reordering, retransmission, and congestion control —
// this transport layer just shuttles opaque IP datagrams.

// Control messages.
const (
	MsgHello    byte = 0x01
	MsgHelloOK  byte = 0x02
	MsgHelloErr byte = 0x03
	MsgPeerConn byte = 0x04
	MsgPeerGone byte = 0x05
	MsgSync     byte = 0x06
	MsgPing     byte = 0xF0
	MsgPong     byte = 0xF1
)

// Data messages.
const (
	// MsgPacket payload = a single raw IP packet (IPv4 or IPv6).
	MsgPacket byte = 0x10
	// MsgPacketBatch payload = repeated [2B lenBE][rawIPpacket].
	// Used by the optional write-coalescing path to reduce per-message overhead.
	MsgPacketBatch byte = 0x11
)

// Frame is a single protocol message.
type Frame struct {
	Type    byte
	Payload []byte
}

// Encode serialises a Frame.
func Encode(f Frame) []byte {
	buf := make([]byte, 1+len(f.Payload))
	buf[0] = f.Type
	copy(buf[1:], f.Payload)
	return buf
}

// Decode parses a byte slice into a Frame.
func Decode(data []byte) (Frame, error) {
	if len(data) < 1 {
		return Frame{}, errors.New("frame too short")
	}
	return Frame{Type: data[0], Payload: data[1:]}, nil
}

// EncodePacketBatch builds a MsgPacketBatch payload from the given packets.
func EncodePacketBatch(packets [][]byte) []byte {
	total := 0
	for _, p := range packets {
		total += 2 + len(p)
	}
	buf := make([]byte, total)
	off := 0
	for _, p := range packets {
		binary.BigEndian.PutUint16(buf[off:], uint16(len(p)))
		off += 2
		copy(buf[off:], p)
		off += len(p)
	}
	return buf
}

// DecodePacketBatch yields each packet to fn. Returns the number of packets
// processed, or an error if the payload is malformed.
func DecodePacketBatch(payload []byte, fn func(pkt []byte)) (int, error) {
	off := 0
	count := 0
	for off < len(payload) {
		if off+2 > len(payload) {
			return count, errors.New("PACKET_BATCH: truncated length prefix")
		}
		l := int(binary.BigEndian.Uint16(payload[off:]))
		off += 2
		if off+l > len(payload) {
			return count, errors.New("PACKET_BATCH: truncated packet")
		}
		fn(payload[off : off+l])
		off += l
		count++
	}
	return count, nil
}

// EncodeHello builds a HELLO payload: [1B version][token UTF-8].
func EncodeHello(version byte, token string) []byte {
	t := []byte(token)
	buf := make([]byte, 1+len(t))
	buf[0] = version
	copy(buf[1:], t)
	return buf
}

// DecodeHello parses a HELLO payload.
func DecodeHello(payload []byte) (version byte, token string, err error) {
	if len(payload) < 1 {
		return 0, "", errors.New("HELLO payload too short")
	}
	return payload[0], string(payload[1:]), nil
}

// EncodeHelloOK builds a HELLO_OK payload:
// [2B ownIdLen][ownId][2B peerIdLen][peerId][2B tokenLen][iamToken]
func EncodeHelloOK(ownID, peerID, iamToken string) []byte {
	o := []byte(ownID)
	p := []byte(peerID)
	t := []byte(iamToken)
	buf := make([]byte, 2+len(o)+2+len(p)+2+len(t))
	off := 0
	binary.BigEndian.PutUint16(buf[off:], uint16(len(o)))
	off += 2
	copy(buf[off:], o)
	off += len(o)
	binary.BigEndian.PutUint16(buf[off:], uint16(len(p)))
	off += 2
	copy(buf[off:], p)
	off += len(p)
	binary.BigEndian.PutUint16(buf[off:], uint16(len(t)))
	off += 2
	copy(buf[off:], t)
	return buf
}

// DecodeHelloOK parses a HELLO_OK payload.
func DecodeHelloOK(payload []byte) (ownID, peerID, iamToken string, err error) {
	if len(payload) < 6 {
		return "", "", "", errors.New("HELLO_OK payload too short")
	}
	off := 0
	oLen := int(binary.BigEndian.Uint16(payload[off:]))
	off += 2
	if off+oLen+2 > len(payload) {
		return "", "", "", errors.New("HELLO_OK: bad ownID length")
	}
	ownID = string(payload[off : off+oLen])
	off += oLen

	pLen := int(binary.BigEndian.Uint16(payload[off:]))
	off += 2
	if off+pLen+2 > len(payload) {
		return "", "", "", errors.New("HELLO_OK: bad peerID length")
	}
	peerID = string(payload[off : off+pLen])
	off += pLen

	tLen := int(binary.BigEndian.Uint16(payload[off:]))
	off += 2
	if off+tLen > len(payload) {
		return "", "", "", errors.New("HELLO_OK: bad token length")
	}
	iamToken = string(payload[off : off+tLen])
	return
}

// EncodePeerConn builds a PEER_CONN payload:
// [2B peerIdLen][peerId][2B tokenLen][iamToken]
func EncodePeerConn(peerID, iamToken string) []byte {
	p := []byte(peerID)
	t := []byte(iamToken)
	buf := make([]byte, 2+len(p)+2+len(t))
	off := 0
	binary.BigEndian.PutUint16(buf[off:], uint16(len(p)))
	off += 2
	copy(buf[off:], p)
	off += len(p)
	binary.BigEndian.PutUint16(buf[off:], uint16(len(t)))
	off += 2
	copy(buf[off:], t)
	return buf
}

// DecodePeerConn parses a PEER_CONN payload.
func DecodePeerConn(payload []byte) (peerID, iamToken string, err error) {
	if len(payload) < 4 {
		return "", "", errors.New("PEER_CONN payload too short")
	}
	off := 0
	pLen := int(binary.BigEndian.Uint16(payload[off:]))
	off += 2
	if off+pLen+2 > len(payload) {
		return "", "", errors.New("PEER_CONN: bad peerID length")
	}
	peerID = string(payload[off : off+pLen])
	off += pLen

	tLen := int(binary.BigEndian.Uint16(payload[off:]))
	off += 2
	if off+tLen > len(payload) {
		return "", "", errors.New("PEER_CONN: bad token length")
	}
	iamToken = string(payload[off : off+tLen])
	return
}

// EncodePong builds a PONG payload: [2B tokenLen][iamToken].
func EncodePong(iamToken string) []byte {
	t := []byte(iamToken)
	buf := make([]byte, 2+len(t))
	binary.BigEndian.PutUint16(buf[0:], uint16(len(t)))
	copy(buf[2:], t)
	return buf
}

// DecodePong parses a PONG payload.
func DecodePong(payload []byte) (iamToken string, err error) {
	if len(payload) < 2 {
		return "", errors.New("PONG payload too short")
	}
	tLen := int(binary.BigEndian.Uint16(payload[0:]))
	if 2+tLen > len(payload) {
		return "", errors.New("PONG: bad token length")
	}
	return string(payload[2 : 2+tLen]), nil
}
